lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,206 @@
+package lib
+
+import (
+	"github.com/dgraph-io/badger/v3"
+	"github.com/pkg/errors"
+	"math"
+	"sort"
+)
+
+type PoSFeeEstimator struct {
+	mempoolFeeEstimatorMempool    *PosMempool
+	pastBlocksFeeEstimatorMempool *PosMempool
+	pastBlocksBadger              *badger.DB
+	numBlocks                     uint64
+	cachedBlocks                  []*MsgDeSoBlock
+	isRunning                     bool
+}
+
+func (posFeeEstimator *PoSFeeEstimator) Init(mempool *PosMempool, pastBlocks []*MsgDeSoBlock, numBlocks uint64) error {
+	if mempool == nil {
+		return errors.New("PoSFeeEstimator.Init: mempool cannot be nil")
+	}
+	if !mempool.IsRunning() {
+		return errors.New("PoSFeeEstimator.Init: mempool must be running")
+	}
+	if len(pastBlocks) == 0 {
+		return errors.New("PoSFeeEstimator.Init: pastBlocks cannot be empty")
+	}
+	if numBlocks == 0 {
+		return errors.New("PoSFeeEstimator.Init: numBlocks cannot be zero")
+	}
+	if numBlocks != uint64(len(pastBlocks)) {
+		return errors.New("PoSFeeEstimator.Init: numBlocks must equal the number of pastBlocks")
+	}
+	// Sort the past blocks by height just to be safe.
+	sort.SliceStable(pastBlocks, func(ii, jj int) bool {
+		return pastBlocks[ii].Header.Height < pastBlocks[jj].Header.Height
+	})
+	minHeight := pastBlocks[0].Header.Height
+	if minHeight == 0 {
+		return errors.New("PoSFeeEstimator.Init: minHeight cannot be zero")
+	}
+	// TODO: Are there any other inputs to validate?
+	posFeeEstimator.mempoolFeeEstimatorMempool = mempool
+	posFeeEstimator.numBlocks = numBlocks
+
+	opts := DefaultBadgerOptions("").WithInMemory(true)
+	var err error
+	if posFeeEstimator.pastBlocksBadger, err = badger.Open(opts); err != nil {
+		return err
+	}
+	pastBlocksUtxoView, err := NewUtxoView(posFeeEstimator.pastBlocksBadger, mempool.params, nil, nil, nil)
+	if err != nil {
+		return errors.Wrap(err, "PoSFeeEstimator.Init: error creating pastBlocksUtxoView")
+	}
+
+	// TODO: Can we get away without setting the pastBlocksUtxoView? This would
+	// save us a badger instance that isn't doing anything.
+
+	// TODO: How do we know when to update global params on the pastBlocksMempool if they're updated on the original mempool?
+	// We only really need that global params entry to help us compute the fee buckets. It's on the caller to update it
+	// using
+	// Create a mempool we can use to estimate fees for past blocks.
+	posFeeEstimator.pastBlocksFeeEstimatorMempool = NewPosMempoolNoValidation(mempool.params, mempool.globalParams.Copy(), pastBlocksUtxoView, minHeight-1, "", true, math.MaxUint64, 0)
+	if err = posFeeEstimator.pastBlocksFeeEstimatorMempool.Start(); err != nil {
+		return errors.Wrap(err, "PosFeeEstimator.Init: error starting pastBlocksMempool")
+	}
```

Comment:
Oof, having to construct a UtxoView just to build an in-memory only mempool is rough. Instead of constructing a mempool for the past blocks estimator, what if you just build a TransactionRegister for the txn? The TransactionRegister is really the data structure we need since that index the txns by txnHash and gives us the fee-time ordering. Worth playing around with it